File: RedeNeural.c
#include <stdlib.h>
#include <math.h>

#define BIAS 1

#define FUNCAO_ATIVACAO(valor) RNA_funcao_ativacao_relu(valor)
#define FUNCAO_ATIVACAO_SAIDA(valor) RNA_funcao_ativacao_reluDx(valor)

typedef struct{
   double* peso;
   double saida;

   int quantidade_ligacoes;
}Neuronio;


typedef struct{
   Neuronio* neuronios;
   int quantidade_neuronios;
}Camada;


typedef struct{
   Camada entrada;
   Camada* ocultas;
   int quantidade_oculta;
   Camada saida;

}RedeNeural;


RedeNeural* RNA_criar_rede(int qtd_neuronios_entrada, int qtd_neuronios_escondida, int qtd_neuronios_saida, int qtd_camadas_escondidas);
void RNA_apagar_rede(RedeNeural* rede);
double RNA_funcao_ativacao_relu(double valor);
double RNA_funcao_ativacao_reluDx(double valor);
double RNA_funcao_ativacao_sigmoid(double valor);
void RNA_inicializar_neuronio(Neuronio* neuronio, int quantidade_ligacoes);
void RNA_imprimir_pesos_entrada(RedeNeural* rede);
void RNA_calcular_entrada(RedeNeural* rede, double dados[]);
void RNA_imprimir_saida_neuronio_saida(RedeNeural* rede);


double RNA_funcao_ativacao_relu(double valor){
   if(valor < 0) return 0;
   else return valor;
}


double RNA_funcao_ativacao_reluDx(double valor){
   if(valor < 0) return 0;
   return 1; 
}


double RNA_funcao_ativacao_sigmoid(double valor){
   return 1 / (1 + exp(-valor));
}


RedeNeural* RNA_criar_rede(int qtd_neuronios_entrada, int qtd_neuronios_ocultas, int qtd_neuronios_saida, int qtd_camadas_ocultas){
   RedeNeural* nova_rede = (RedeNeural*) malloc(sizeof(RedeNeural));

   //camadas de entrada
   nova_rede->entrada.quantidade_neuronios = qtd_neuronios_entrada;
   nova_rede->entrada.neuronios = (Neuronio*) malloc(qtd_neuronios_entrada * sizeof(Neuronio));
   for(int i = 0; i < qtd_neuronios_entrada; i++){
      RNA_inicializar_neuronio(&nova_rede->entrada.neuronios[i], qtd_neuronios_ocultas);
   }

   //camadas ocultas
   nova_rede->ocultas = (Camada*) malloc(qtd_camadas_ocultas * sizeof(Camada));
   for(int i = 0; i < qtd_camadas_ocultas; i++){
      nova_rede->ocultas[i].quantidade_neuronios = qtd_neuronios_ocultas;
      nova_rede->ocultas[i].neuronios = (Neuronio*) malloc(qtd_neuronios_ocultas * sizeof(Neuronio));


      for(int j = 0; j < qtd_neuronios_ocultas; j++){
         if(i == qtd_camadas_ocultas-1){
            RNA_inicializar_neuronio(&nova_rede->ocultas[i].neuronios[j], qtd_neuronios_saida);
         }else{
            RNA_inicializar_neuronio(&nova_rede->ocultas[i].neuronios[j], qtd_neuronios_ocultas);
         }
      }
   }
   nova_rede->quantidade_oculta = qtd_camadas_ocultas;


   //cada de saída
   nova_rede->saida.quantidade_neuronios = qtd_neuronios_saida;
   nova_rede->saida.neuronios = (Neuronio*) malloc(qtd_neuronios_saida * sizeof(Neuronio));
   for(int i = 0; i < qtd_neuronios_saida; i++){
      RNA_inicializar_neuronio(&nova_rede->saida.neuronios[i], qtd_neuronios_saida);
   }

   return nova_rede;
}


void RNA_apagar_rede(RedeNeural* rede){
   free(rede->entrada.neuronios);
   for(int i = 0; i < rede->quantidade_oculta; i++){
      free(rede->ocultas[i].neuronios);
   }
   free(rede->saida.neuronios);
   free(rede->ocultas);
   free(rede);   
}


void RNA_inicializar_neuronio(Neuronio* neuronio, int quantidade_ligacoes){
   neuronio->quantidade_ligacoes = quantidade_ligacoes;
   neuronio->peso = (double*) malloc(quantidade_ligacoes * sizeof(double));

   for(int i = 0; i < quantidade_ligacoes; i++){
      neuronio->peso[i] = (rand() % 200) - 100;
   }

   neuronio->saida = 0;
}


void RNA_calcular_entrada(RedeNeural* rede, double dados[]){

   //camada de entrada
   int qtd_neuronios_entrada = rede->entrada.quantidade_neuronios;
   double soma = 0;
   for(int i = 0; i < qtd_neuronios_entrada; i++){
      rede->entrada.neuronios[i].saida = dados[i];
   }

   //camadas ocultas
   for (int camada_oculta = 0; camada_oculta < rede->quantidade_oculta; camada_oculta++){
      /*referencias para as camadas atual e anterior
       * Atualiza as referencias com base na camada oculta atual
      */
      Camada* camada_atual = &(rede->ocultas[camada_oculta]);
      Camada* camada_anterior;
      if (camada_oculta == 0) camada_anterior = &(rede->entrada);
      else camada_anterior = &(rede->ocultas[camada_oculta - 1]);

      /*percorre os neuronios da camada atual
       * e para cada neuronio da camada atual, faz a soma
       * por todos os neuronios da camada anterior
      */ 
      for(int neuronio = 0; neuronio < (camada_atual->quantidade_neuronios); neuronio++){
         double soma = 0;

         for (int i = 0; i < (camada_anterior->quantidade_neuronios); i++){
            soma += camada_anterior->neuronios[i].saida * camada_atual->neuronios[neuronio].peso[i];
         }
         soma += BIAS;
         camada_atual->neuronios[neuronio].saida = FUNCAO_ATIVACAO(soma);
      }
   }

   //camada de saída
   //percorrer cada neuronio de saída
   for(int neuronio_saida = 0; neuronio_saida < (rede->saida.quantidade_neuronios); neuronio_saida++){
      Camada* oculta = &rede->ocultas[rede->quantidade_oculta-1];

      //percorrer cada neuronio da camada oculta anterior
      double soma = 0;
      for(int neuronio_oculta = 0; neuronio_oculta < oculta->quantidade_neuronios; neuronio_oculta++){
         soma += (oculta->neuronios[neuronio_oculta].saida) * (rede->saida.neuronios[neuronio_saida].peso[neuronio_oculta]);
      }
      soma += BIAS;
      rede->saida.neuronios[neuronio_saida].saida = FUNCAO_ATIVACAO_SAIDA(soma);
   }
}


void RNA_imprimir_saidas_entrada(RedeNeural* rede){
   printf("--Entrada--\n");
   for(int i = 0; i < rede->entrada.quantidade_neuronios; i++){
      printf("[neuronio(%d): %.2f]\n", i, rede->entrada.neuronios[i].saida);
   }
   printf("\n"); 
}


void RNA_imprimir_saidas_ocultas(RedeNeural* rede){
   printf("--Camadas ocultas--\n");
   for(int camada_escondida = 0; camada_escondida < (rede->quantidade_oculta); camada_escondida++){
      printf("Camada %d\n", camada_escondida);
      for(int i = 0; i < rede->ocultas[camada_escondida].quantidade_neuronios; i++){
         printf("[neuronio(%d): %.2f]\n", i, rede->ocultas[camada_escondida].neuronios[i].saida);
      }
   }
   printf("\n");
}


void RNA_imprimir_saidas_saida(RedeNeural* rede){
   printf("--Saida--\n");
   for(int i = 0; i < rede->saida.quantidade_neuronios; i++){
      printf("[neuronio(%d): %.2f]\n", i, rede->saida.neuronios[i].saida);
   }
   printf("\n");
}

File: Teste RNA/rna/RedeNeural.java
package rna;


public class RedeNeural implements Cloneable{
   public Camada entrada;
   public Camada[] ocultas;
   public Camada saida;
   
   public int qtdNeuroniosEntrada;
   public int qtdNeuroniosOcultas;
   public int qtdNeuroniosSaida;

   public int qtdCamadasOcultas;

   int BIAS = 1;
   double TAXA_APRENDIZAGEM = 0.1;

   //padronizar uso das funções de ativação
   final int ativacaoRelu = 1;
   final int ativacaoReluDx = 2;
   final int ativacaoSigmoid = 3;
   final int ativacaoSigmoidDx = 4;
   final int ativacaoTanH = 5;
   
   int funcaoAtivacao = ativacaoTanH;
   int funcaoAtivacaoSaida = ativacaoTanH;

   public RedeNeural(int qtdNeuroniosEntrada, int qtdNeuroniosOcultas, int qtdNeuroniosSaida, int qtdCamadasOcultas){
      if(qtdNeuroniosEntrada < 1 || qtdNeuroniosOcultas < 1 || qtdNeuroniosSaida < 1 || qtdCamadasOcultas < 1){
         throw new IllegalArgumentException("Os valores devem ser maiores ou iguais a um.");
      }

      this.qtdNeuroniosEntrada = qtdNeuroniosEntrada;
      this.qtdNeuroniosOcultas = qtdNeuroniosOcultas;
      this.qtdNeuroniosSaida = qtdNeuroniosSaida;

      this.qtdCamadasOcultas = qtdCamadasOcultas;

      criarRede();
   }


   private void criarRede(){
      //inicializar camada de entrada
      entrada = new Camada();
      entrada.neuronios = new Neuronio[qtdNeuroniosEntrada];
      for(int i = 0; i < qtdNeuroniosEntrada; i++){
         entrada.neuronios[i] = new Neuronio(qtdNeuroniosOcultas);
      }

      //inicializar camadas ocultas
      ocultas = new Camada[qtdCamadasOcultas];
      for (int i = 0; i < qtdCamadasOcultas; i++) {
         Camada novaOculta = new Camada();
         novaOculta.neuronios = new Neuronio[qtdNeuroniosOcultas];
      
         for (int j = 0; j < qtdNeuroniosOcultas; j++) {
            if (i == (qtdCamadasOcultas-1)){
               novaOculta.neuronios[j] = new Neuronio(qtdNeuroniosSaida);
            
            }else{
               novaOculta.neuronios[j] = new Neuronio(qtdNeuroniosOcultas);
            }
         }
         ocultas[i] = novaOculta;
      }

      //inicializar camada de saída
      saida = new Camada();
      saida.neuronios = new Neuronio[qtdNeuroniosSaida];
      for(int i = 0; i < qtdNeuroniosSaida; i++){
         saida.neuronios[i] = new Neuronio(qtdNeuroniosSaida);
      }

   }


   public void calcularSaida(RedeNeural rede, double[] dados){
      if(dados.length != rede.entrada.neuronios.length){
         throw new IllegalArgumentException("As dimensões dos dados de entrada com os neurônios de entrada da rede não são iguais");
      }

      //entrada
      for(int i = 0; i < rede.qtdNeuroniosEntrada; i++){
         rede.entrada.neuronios[i].saida = dados[i];
      }

      //ocultas
      double soma = 0.0;
      for(int i = 0; i < rede.qtdCamadasOcultas; i++){//percorrer camadas ocultas

         Camada camadaAtual = rede.ocultas[i];
         Camada camadaAnterior;
         if(i == 0) camadaAnterior = rede.entrada;
         else camadaAnterior = rede.ocultas[i-1];

         for(int j = 0; j < camadaAtual.neuronios.length; j++){
            soma = 0.0;
            for(int k = 0; k < camadaAnterior.neuronios.length; k++){
               soma += camadaAnterior.neuronios[k].saida * camadaAnterior.neuronios[k].pesos[j];
            }
            soma += BIAS;
            camadaAtual.neuronios[j].saida = funcaoAtivacao(soma);
         }
      }

      //saída
      for(int neuronioSaida = 0; neuronioSaida < rede.saida.neuronios.length; neuronioSaida++){
         Camada oculta = rede.ocultas[rede.qtdCamadasOcultas-1];
         soma = 0.0;

         for(int neuronioOculta = 0; neuronioOculta < oculta.neuronios.length; neuronioOculta++){
            soma += oculta.neuronios[neuronioOculta].saida * oculta.neuronios[neuronioOculta].pesos[neuronioSaida]; 
         }
         soma += BIAS;
         rede.saida.neuronios[neuronioSaida].saida = funcaoAtivacaoSaida(soma);
      }
   }

   
   public double calcularPrecisao(RedeNeural rede, double[][] dados){
      double precisao = 0;

      double[] dados_treino = new double[dados[0].length-1];
      double[] classe_treino = new double[1];
      int acertosTotais = 0;
      int acertosSaida = 0;

      for(int i = 0; i < dados.length; i++){
         dados_treino[0] = dados[i][0];
         dados_treino[1] = dados[i][1];
         dados_treino[2] = dados[i][2];
         classe_treino[0]= dados[i][3];

         rede.calcularSaida(rede, dados_treino);

         for(int j = 0; j < rede.saida.neuronios.length; j++){
            if(rede.saida.neuronios[j].saida == classe_treino[j]){
               acertosSaida ++;
            }
         }
         
         if(acertosSaida == (rede.saida.neuronios.length)) acertosTotais++;

      }

      precisao = (double) acertosTotais/dados.length;
      return (precisao*100);
   }


   //funções de ativação
   /**
    * 1 - ReLu
    * 2 - ReLu derivada
    * 3 - Sigmoide
    * 4 - Sigmoid derivada
    * 5 - Tangente tangente hiperbolica
    * Por padrão será usado ReLu e Relu derivada, respectivamente
    * @param ocultas função de ativação das camadas ocultas
    * @param saida função de ativação da ultima camada oculta para a saída
    */
   public void configurarFuncaoAtivacao(int ocultas, int saida){
      funcaoAtivacao = ocultas;
      funcaoAtivacaoSaida = saida;
   }


   public double funcaoAtivacao(double valor){
      if(funcaoAtivacao == ativacaoRelu) return relu(valor);
      if(funcaoAtivacao == ativacaoReluDx) return reluDx(valor);
      if(funcaoAtivacao == ativacaoSigmoid) return sigmoid(valor);
      if(funcaoAtivacao == ativacaoSigmoidDx) return sigmoidDx(valor);
      if(funcaoAtivacao == ativacaoTanH) return tanH(valor);

      else return valor;
   }


   public double funcaoAtivacaoSaida(double valor){
      if(funcaoAtivacaoSaida == ativacaoRelu) return relu(valor);
      if(funcaoAtivacaoSaida == ativacaoReluDx) return reluDx(valor);
      if(funcaoAtivacaoSaida == ativacaoSigmoid) return sigmoid(valor);
      if(funcaoAtivacaoSaida == ativacaoSigmoidDx) return sigmoidDx(valor);
      if(funcaoAtivacaoSaida == ativacaoTanH) return tanH(valor);

      else return valor;
   }


   private double relu(double valor){
      if(valor < 0) return 0;
      return valor;
   }


   private double reluDx(double valor){
      if(valor < 0) return 0;
      return 1;     
   }


   private double sigmoid(double valor){
      return 1 / (1 + Math.exp(-valor));
   }


   private double sigmoidDx(double valor){
      return (sigmoid(valor) * (1-sigmoid(valor)));
   }


   private double tanH(double valor){
      return Math.tanh(valor);
   }


   //clonar a rede
   @Override
   public RedeNeural clone(){
      try {
         RedeNeural clone = (RedeNeural) super.clone();

         // Clonar dados importantes
         clone.qtdNeuroniosEntrada = this.qtdNeuroniosEntrada;
         clone.qtdNeuroniosOcultas = this.qtdNeuroniosOcultas;
         clone.qtdNeuroniosSaida = this.qtdNeuroniosSaida;
         clone.qtdCamadasOcultas = this.qtdCamadasOcultas;
         clone.BIAS = this.BIAS;
         clone.TAXA_APRENDIZAGEM = this.TAXA_APRENDIZAGEM;

         // Clonar camada de entrada
         clone.entrada = cloneCamada(this.entrada);

         // Clonar camadas ocultas
         clone.ocultas = new Camada[qtdCamadasOcultas];
         for (int i = 0; i < qtdCamadasOcultas; i++) {
            clone.ocultas[i] = cloneCamada(this.ocultas[i]);
         }

         // Clonar camada de saída
         clone.saida = cloneCamada(this.saida);

         return clone;
      } catch (CloneNotSupportedException e){
         throw new RuntimeException(e);
      }
   }


   private Camada cloneCamada(Camada camada){
      Camada clone = new Camada();
      clone.neuronios = new Neuronio[camada.neuronios.length];

      for (int i = 0; i < camada.neuronios.length; i++) {
         clone.neuronios[i] = cloneNeuronio(camada.neuronios[i], camada.neuronios[i].qtdLigacoes, camada.neuronios[i].pesos);
      }

      return clone;
   }


   private Neuronio cloneNeuronio(Neuronio neuronio, int qtdLigacoes, double[] pesos){
      Neuronio clone = new Neuronio(neuronio.pesos.length);

      double pesosClone[] = new double[qtdLigacoes];

      for(int i = 0; i < pesos.length; i++){
         pesosClone[i] = pesos[i];
      }

      clone.pesos = pesosClone;
      clone.qtdLigacoes = qtdLigacoes;

      return clone;
   }
}
